kiosk/sources/extensions/collection_bidding_ext.move

```diff
@@ -0,0 +1,220 @@
+        ext::add(Extension {}, self, cap, PERMISSIONS, ctx)
+    }
+
+    // === Bidding logic ===
+
+    /// Place a bid on any item in a collection (`T`). We do not assert that all
+    /// the values in the `place_bids` are identical, the amounts are emitted
+    /// in the event, the order is reversed.
+    ///
+    /// Use `sui::pay::split_n` to prepare the Coins for the bid.
+    public fun place_bids<T: key + store, Market>(
+        self: &mut Kiosk, cap: &KioskOwnerCap, bids: vector<Coin<SUI>>, _ctx: &mut TxContext
+    ) {
+        assert!(vector::length(&bids) > 0, ENoCoinsPassed);
+        assert!(kiosk::has_access(self, cap), ENotAuthorized);
+
+        let amounts = vector[];
+        let (i, count) = (0, vector::length(&bids));
+        while (i < count) {
+            vector::push_back(&mut amounts, coin::value(vector::borrow(&bids, i)));
+            i = i + 1;
+        };
+
+        event::emit(NewBid<T, Market> {
+            kiosk_id: object::id(self),
+            bids: amounts,
+            kiosk_owner: personal_kiosk::try_owner(self)
+        });
+
+        bag::add(ext::storage_mut(Extension {}, self), Bid<T, Market> {}, bids);
+    }
+
+    /// Cancel all bids, return the funds to the owner.
+    public fun cancel_all<T: key + store, Market>(
+        self: &mut Kiosk, cap: &KioskOwnerCap, ctx: &mut TxContext
+    ): Coin<SUI> {
+        assert!(kiosk::has_access(self, cap), ENotAuthorized);
+
+        event::emit(BidCanceled<T, Market> {
+            kiosk_id: object::id(self),
+            kiosk_owner: personal_kiosk::try_owner(self)
+        });
+
+        let coins = bag::remove(ext::storage_mut(Extension {}, self), Bid<T, Market> {});
+        let total = coin::zero(ctx);
+        pay::join_vec(&mut total, coins);
+        total
+    }
+
+    /// Accept the bid and make a purchase on in the `Kiosk`.
+    ///
+    /// 1. The seller creates a `MarketPurchaseCap` using the Marketplace adapter,
+    /// and passes the Cap to this function.
+    ///
+    /// 2. The `bid` is taken from the extension storage and is used to purchase
+    /// the item with the `MarketPurchaseCap`. Proceeds go to the seller's Kiosk.
+    ///
+    /// 3. The item is placed in the seller's Kiosk using the `place` or `lock`
+    /// functions (see `PERMISSIONS`).
+    public fun accept_market_bid<T: key + store, Market>(
+        destination: &mut Kiosk,
+        source: &mut Kiosk,
+        mkt_cap: MarketPurchaseCap<T, Market>,
+        policy: &TransferPolicy<T>,
+        // keeping these arguments for extendability
+        _lock: bool,
+        ctx: &mut TxContext
+    ): (TransferRequest<T>, TransferRequest<Market>) {
+        let storage = ext::storage_mut(Extension {}, destination);
+        assert!(bag::contains(storage, Bid<T, Market> {}), EBidNotFound);
+
+        // Take 1 Coin from the bag - this is our bid (bids can't be empty, we make sure of it).
+        let bid = vector::pop_back(bag::borrow_mut(storage, Bid<T, Market> {}));
```

Comment:
As someone who wants to buy multiple Capys I can submit multiple bids, eg:

10 SUI, 7 SUI, 5 SUI

So that whoever wants to accept my offer can take 10 SUI in exchange for a Capy first. Then the next "seller" gets 7 SUI and so on. It's usually done in batches, like on BlueMove, for instance, you specify the amount of offers you want to place: like "10 SUI, 5 offers".